api/debug.go

```diff
@@ -22,77 +38,392 @@ type txTraceResult struct {
+func NewDebugAPI(
+	store *pebble.Storage,
+	registerStore *pebble.RegisterStorage,
+	tracer storage.TraceIndexer,
+	blocks storage.BlockIndexer,
+	transactions storage.TransactionIndexer,
+	receipts storage.ReceiptIndexer,
+	config *config.Config,
+	logger zerolog.Logger,
+	collector metrics.Collector,
+) *DebugAPI {
 	return &DebugAPI{
-		logger:    logger,
-		tracer:    tracer,
-		blocks:    blocks,
-		collector: collector,
+		store:         store,
+		registerStore: registerStore,
+		logger:        logger,
+		tracer:        tracer,
+		blocks:        blocks,
+		transactions:  transactions,
+		receipts:      receipts,
+		config:        config,
+		collector:     collector,
 	}
 }
 
 // TraceTransaction will return a debug execution trace of a transaction if it exists,
 // currently we only support CALL traces, so the config is ignored.
 func (d *DebugAPI) TraceTransaction(
-	_ context.Context,
+	ctx context.Context,
 	hash gethCommon.Hash,
-	_ *tracers.TraceConfig,
+	config *tracers.TraceConfig,
 ) (json.RawMessage, error) {
-	res, err := d.tracer.GetTransaction(hash)
+	// If the given trace config is equal to the default call tracer used
+	// in block replay during ingestion, then we fetch the trace result
+	// from the Traces DB.
+	if isDefaultCallTracer(config) {
+		trace, err := d.tracer.GetTransaction(hash)
+		if err == nil {
+			return trace, nil
+		}
+	}
+
+	receipt, err := d.receipts.GetByTransactionID(hash)
+	if err != nil {
+		return nil, err
+	}
+
+	tracer, err := tracerForReceipt(config, receipt)
+	if err != nil {
+		return nil, err
+	}
+
+	block, err := d.blocks.GetByHeight(receipt.BlockNumber.Uint64())
+	if err != nil {
+		return nil, err
+	}
+	// We need to re-execute the given transaction and all the
+	// transactions that precede it in the same block, based on
+	// the previous block state, to generate the correct trace.
+	previousBlock, err := d.blocks.GetByHeight(block.Height - 1)
+	if err != nil {
+		return nil, err
+	}
+
+	blockExecutor, err := d.executorAtBlock(previousBlock)
 	if err != nil {
-		return handleError[json.RawMessage](err, d.logger, d.collector)
+		return nil, err
 	}
-	return res, nil
+
+	// Re-execute the transactions in the order they appear, for the block
+	// that contains the given transaction. We set the tracer only for
+	// the given transaction, as we don't need it for the preceding
+	// transactions. Once we re-execute the desired transaction, we ignore
+	// the rest of the transactions in the block, and simply return the trace
+	// result.
+	txExecuted := false
+	var txTracer *tracers.Tracer
+	for _, h := range block.TransactionHashes {
+		if txExecuted {
+			break
+		}
+
+		tx, err := d.transactions.Get(h)
+		if err != nil {
+			return nil, err
+		}
+
+		if h == hash {
+			txTracer = tracer
+			txExecuted = true
+		}
+
+		_, err = blockExecutor.Run(tx, txTracer)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return txTracer.GetResult()
```

Comment:
tx trace could theoretically be nil, I think lets just cover that code path with an error.

Reply:
Good point :+1: Added a `nil` check in https://github.com/onflow/flow-evm-gateway/pull/635/commits/3c6a4ff81be9f67f1580d27c057288209c9acca6